src/stream/src/common/table/state_table.rs

```diff
@@ -57,6 +58,63 @@ use crate::executor::{StreamExecutorError, StreamExecutorResult};
 /// This num is arbitrary and we may want to improve this choice in the future.
 const STATE_CLEANING_PERIOD_EPOCH: usize = 5;
 
+/// Used for watermark state cleaning.
+/// We store the smallest value in our state table here.
+///
+/// Cases
+/// -----
+/// Uninitialized -> No value in cache, because none inserted yet. No delete range required.
+/// Empty -> No value in cache. We need to issue delete range,
+///          since we don't know the smallest value.
+/// Smallest(value) if `current_watermark` < value -> No delete range required.
+/// Smallest(value) if `current_watermark` >= value -> Delete range (-inf, `current_watermark`].
+///
+/// Updating the Cache
+/// ------------------
+/// If no watermark column, just set watermark state to Empty.
+/// Otherwise if watermark column exists:
+/// (UPDATE)INSERT -> Update lowest value if applicable.
+/// (UPDATE)DELETE -> If removes lowest value, set cache to empty.
+#[derive(Clone, PartialEq, Eq, Debug)]
+enum WatermarkCache {
+    Empty,
+    Uninitialized,
+    // We need to store row, in case we encounter `Delete`s.
+    // Then we can remove the OwnedRow.
+    Smallest(OwnedRow),
+}
+
+impl WatermarkCache {
+    /// Handle inserts / updates.
+    fn handle_insert(&mut self, row: impl Row, watermark_col_idx: &Option<usize>) {
+        let Some(idx) = watermark_col_idx else {
+            *self = Self::Empty;
+            return;
+        };
+        let new_value = row.datum_at(*idx);
+        match self {
+            Self::Smallest(current_row) => {
+                let current_value = current_row.datum_at(*idx);
+                if cmp_datum(new_value, current_value.to_datum_ref(), Default::default())
+                    == Ordering::Less
+                {
+                    *self = Self::Smallest(row.into_owned_row());
+                }
+                // Otherwise keep the old value.
+            }
+            Self::Uninitialized => {
+                *self = Self::Smallest(row.into_owned_row());
+            }
+            Self::Empty => {
+                // Empty -> We don't know the smallest value.
+                // So we can't update anything.
+                // We have to do a table scan
+                // if we wish to refresh the cache.
```

Comment:
Either we need to do:
1. Table scan on barrier.
2. Table scan on Delete.

Or is there anyway to avoid this?

Reply:
I will opt for table scan on Delete for now.
We can optimize it later if needed.